exercises/practice/forth/test/tests.chpl

```diff
@@ -0,0 +1,292 @@
+  test.assertEqual(evaluate(["1 2 3 over"]), [1, 2, 3, 2]);
+}
+
+proc testOverErrorsIfThereIsNothingOnTheStack(test : borrowed Test) throws {
+  try {
+    evaluate(["over"]);
+    throw new owned AssertionError("function was expected to throw an error");
+  } catch e: IllegalArgumentError {
+    test.assertEqual(e.message(), "empty stack");
+  }
+}
+
+proc testOverErrorsIfThereIsOnlyOneValueOnTheStack(test : borrowed Test) throws {
+  try {
+    evaluate(["1 over"]);
+    throw new owned AssertionError("function was expected to throw an error");
+  } catch e: IllegalArgumentError {
+    test.assertEqual(e.message(), "only one value on the stack");
+  }
+}
+
+proc testUserDefinedWordsCanConsistOfBuiltInWords(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": dup-twice dup dup ;", "1 dup-twice"]), [1, 1, 1]);
+}
+
+proc testUserDefinedWordsExecuteInTheRightOrder(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": countup 1 2 3 ;", "countup"]), [1, 2, 3]);
+}
+
+proc testUserDefinedWordsCanOverrideOtherUserDefinedWords(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": foo dup ;", ": foo dup dup ;", "1 foo"]), [1, 1, 1]);
+}
+
+proc testUserDefinedWordsCanOverrideBuiltInWords(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": swap dup ;", "1 swap"]), [1, 1]);
+}
+
+proc testUserDefinedWordsCanOverrideBuiltInOperators(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": + * ;", "3 4 +"]), [12]);
+}
+
+proc testUserDefinedWordsCanUseDifferentWordsWithTheSameName(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": foo 5 ;", ": bar foo ;", ": foo 6 ;", "bar foo"]), [5, 6]);
+}
+
+proc testUserDefinedWordsCanDefineWordThatUsesWordWithTheSameName(test : borrowed Test) throws {
+  test.assertEqual(evaluate([": foo 10 ;", ": foo foo 1 + ;", "foo"]), [11]);
+}
+
+proc testUserDefinedWordsCannotRedefineNonNegativeNumbers(test : borrowed Test) throws {
+  try {
+    evaluate([": 1 2 ;"]);
+    throw new owned AssertionError("function was expected to throw an error");
+  } catch e: IllegalArgumentError {
+    test.assertEqual(e.message(), "illegal operation");
+  }
+}
+
+proc testUserDefinedWordsCannotRedefineNegativeNumbers(test : borrowed Test) throws {
+  try {
+    evaluate([": -1 2 ;"]);
+    throw new owned AssertionError("function was expected to throw an error");
+  } catch e: IllegalArgumentError {
+    test.assertEqual(e.message(), "illegal operation");
+  }
+}
+
+proc testUserDefinedWordsErrorsIfExecutingANonExistentWord(test : borrowed Test) throws {
+  try {
+    evaluate(["foo"]);
+    throw new owned AssertionError("function was expected to throw an error");
+  } catch e: IllegalArgumentError {
+    test.assertEqual(e.message(), "undefined operation");
+  }
+}
+
+proc testUserDefinedWordsOnlyDefinesLocally(test : borrowed Test) throws {
+  test.assertEqual(evaluateBoth([[": + - ;", "1 1 +"], {{instructionsSecond}}), [[0], [2]]);
```

Comment:
I noticed that I had forgotten to properly define this test. I will fix it. If you start working on this in the meanwhile, just delete the test, and I'll add it back when I've figured it out.

Reply:
good catch! The past few weeks have been very busy, but I hope I'll manage to give the final push to this in the upcoming days... it's not far from launching!